### ISSUE TITLE
Does cudnn Batchnorm support 5D input

### ISSUE THREAD

**Issue by ascust** [apache/incubator-mxnet#5211]
I am trying to use 3D convolution as well as 3D batch norm, which means their input are 5D tensors. There is nothing wrong with conv layer, but I got an error when using cudnn batch norm. I heard 5D is supported by cudnn bn, but why I got this error? Something wrong with the mxnet implementation?

**Comment by mli** [comment]
what's error msg you got?

batchnorm should works for general N-D input: http://ec2-54-152-159-175.compute-1.amazonaws.com///api/python/symbol.html#mxnet.symbol.BatchNorm

**Comment by ascust** [comment]
I got:
[10:22:55] /home/a1689670/mxnet/dmlc-core/include/dmlc/./logging.h:300: [10:22:55] src/engine/./threaded_engine.h:336: [10:22:55] src/operator/./cudnn_batch_norm-inl.h:59: Check failed: in_data[cudnnbatchnorm::kData].ndim() <= 4 (5 vs. 4) 
@mli

**Comment by ascust** [comment]
Any help please? @piiswrong @tqchen @pluskid

**Comment by pluskid** [comment]
It seems the current cudnn batchnorm implementation requires the input dimension <= 4. Updates are needed to either call cuDNN's general ND tensor API or use `get_with_shape` to fake 4D tensors out of ND.

https://github.com/dmlc/mxnet/blob/master/src/operator/cudnn_batch_norm-inl.h#L62

**Comment by leezu** [comment]
For now a workaround is to reshape the 5D data to 4D, perform the BatchNorm and then reshape back to 5D.

You may use the following code:

```
def batchnorm_5d(data, height, width, name, fix_gamma, eps, momentum):
    data = mx.symbol.reshape(data, shape=(0, 0, -1, width))

    data = BatchNorm(
        data,
        name=name,
        fix_gamma=fix_gamma,
        eps=eps,
        momentum=momentum)

    return mx.symbol.reshape(data, shape=(0, 0, -1, height, width))
```

This should be equivalent to the way cudnn performs the 5D (spatial) BatchNorm:

> CUDNN_BATCHNORM_SPATIAL: Normalization is performed over N+spatial
> dimensions. This mode is intended for use after
> convolutional layers (where spatial invariance
> is desired). In this mode bnBias, bnScale tensor
> dimensions are 1xCx1x1.

As cudnn throws `CUDNN_STATUS_BAD_PARAM` if:

> bnScaleBiasMeanVarDesc dimensions are not
> 1xC(x1)x1x1 for spatial or 1xC(xD)xHxW for
> per-activation mode (parenthesis for 5D).

(See the user manual at https://developer.nvidia.com/cudnn )

**Comment by jovapo** [comment]
Has anyone made any progress on this? 

The workaround posted will make network creation quite rigid as you need to know the input size and compute the current input size while constructing the network. 

Is there any way to compute height and width dynamically so they don't need to be passed in? Also is there any issue with performance with the above workaround?

**Comment by leezu** [comment]
You can refer to the formulas in https://arxiv.org/abs/1603.07285 and write some helper functions to calculate the shape automatically. Alternatively you can call the `infer_shape` method on a symbol, passing the input data shapes as keyword arguments. It will return a list of 3 elements, with the output shape being the second one. Based on that you know the arguments for the 5d_batchnorm function. This could be done automatically when constructing your symbols.

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
